False positive used-before-assignment message when using assignment expression as a function argument

### Bug description

```python
# .vscode/scratch/scratch.py

def my_func1(candidate):
    return candidate

def my_func2(my_dict):
    return my_func1(
        candidate=candidate if (candidate := my_dict.get("path")) else None
    )
```


### Configuration

_No response_

### Command used

```shell
python3.8 -m pylint .vscode/scratch/scratch.py
```


### Pylint output

```shell
************* Module scratch
.vscode/scratch/scratch.py:8:18: E0601: Using variable 'candidate' before assignment (used-before-assignment)

--------------------------------------------------------------------
Your code has been rated at -2.50/10 (previous run: -2.50/10, +0.00)
```


### Expected behavior

Would not expect any errors, since the assingment expression runs first. Interestingly, assigning the expression to a variable first does not generate the same error. The following code checks cleanly.

```
# .vscode/scratch/scratch.py

def my_func1(candidate):
    return candidate

def my_func2(my_dict):
    candidate = candidate if (candidate := my_dict.get("path")) else None
    return my_func1(
        candidate=candidate
    )

```

### Pylint version

```shell
pylint 2.11.1
astroid 2.8.3
Python 3.8.0 (default, Oct 28 2019, 16:14:01) 
[GCC 8.3.0]
```


### OS / Environment

NAME="Ubuntu"
VERSION="18.04.2 LTS (Bionic Beaver)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 18.04.2 LTS"
VERSION_ID="18.04"
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
VERSION_CODENAME=bionic
UBUNTU_CODENAME=bionic

### Additional dependencies

_No response_

--- a/ChangeLog
+++ b/ChangeLog
@@ -102,6 +102,12 @@ Release date: TBA
 
   Closes #5569
 
+* Fix false positives for ``used-before-assignment`` from using named
+  expressions in a ternary operator test and using that expression as
+  a call argument.
+
+  Closes #5177, #5212
+
 * Fix false positive for ``undefined-variable`` when ``namedtuple`` class
   attributes are used as return annotations.
 

--- a/doc/whatsnew/2.13.rst
+++ b/doc/whatsnew/2.13.rst
@@ -154,6 +154,12 @@ Other Changes
 
   Closes #4941
 
+* Fix false positives for ``used-before-assignment`` from using named
+  expressions in a ternary operator test and using that expression as
+  a call argument.
+
+  Closes #5177, #5212
+
 * Fix ``unnecessary_dict_index_lookup`` false positive when deleting a dictionary's entry.
 
   Closes #4716

--- a/pylint/checkers/variables.py
+++ b/pylint/checkers/variables.py
@@ -1965,8 +1965,23 @@ class VariablesChecker(BaseChecker):
                     )
                     and (
                         isinstance(defstmt.value, nodes.IfExp)
-                        or isinstance(defstmt.value, nodes.Lambda)
-                        and isinstance(defstmt.value.body, nodes.IfExp)
+                        or (
+                            isinstance(defstmt.value, nodes.Lambda)
+                            and isinstance(defstmt.value.body, nodes.IfExp)
+                        )
+                        or (
+                            isinstance(defstmt.value, nodes.Call)
+                            and (
+                                any(
+                                    isinstance(kwarg.value, nodes.IfExp)
+                                    for kwarg in defstmt.value.keywords
+                                )
+                                or any(
+                                    isinstance(arg, nodes.IfExp)
+                                    for arg in defstmt.value.args
+                                )
+                            )
+                        )
                     )
                     and frame is defframe
                     and defframe.parent_of(node)

--- a/tests/functional/u/undefined/undefined_variable_py38.py
+++ b/tests/functional/u/undefined/undefined_variable_py38.py
@@ -140,3 +140,19 @@ def type_annotation_used_improperly_after_comprehension_2():
     my_int: int
     _ = [print(my_int, my_int) for my_int in range(10)]
     print(my_int)  # [used-before-assignment]
+
+
+# Tests for named expressions (walrus operator)
+
+# Expression in ternary operator: positional argument
+print(sep=colon if (colon := ":") else None)
+
+
+class Dummy:
+    """Expression in ternary operator: keyword argument"""
+    # pylint: disable=too-few-public-methods
+    def __init__(self, value):
+        self.value = value
+
+
+dummy = Dummy(value=val if (val := 'something') else 'anything')

